src/Lib.hs

```diff
@@ -0,0 +1,72 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Lib
+    (
+     ascii_art_str,
+     basic_str,
+     basic_const_zero,
+     BasicExternal,
+    ) where
+
+import Foreign.C.Types
+import Foreign.Ptr
+import Foreign.C.String
+import Foreign.Storable
+import Foreign.Marshal.Array
+import Foreign.ForeignPtr
+import Control.Applicative
+
+
+data BasicStruct = BasicStruct {
+    data_ptr :: Ptr ()
+}
+
+instance Storable BasicStruct where
+    alignment _ = 8
+    sizeOf _ = sizeOf nullPtr
+    peek basic_ptr = BasicStruct <$> peekByteOff basic_ptr 0
+    poke basic_ptr BasicStruct{..} = pokeByteOff basic_ptr 0 data_ptr
+
+
+type BasicExternal = ForeignPtr BasicStruct
+-- !an array of size 1 of type `BasicStruct`
+type BasicInternal = Ptr BasicStruct
```

Comment:
a `Ptr` is just a raw FFI pointer, while a `ForeignPtr` is a "tracked" FFI pointer (in the sense of memory deallocation). I create a `ForeignPtr` with the `newForeignPtr` function that couples the raw pointer with a memory deallocation function pointer